### DISCUSSION TITLE
Problems with the PropertyGrid control

### DISCUSSION THREAD

**Discussion by hap1983** [IronLanguages/ironpython3#1633] [Q&A]
When using the PropertyGrid control in ironpython, how to use the CategoryAttribute feature to classify and display properties in the PropertyGrid control?
I know how to do it in C#，using Attribute label：
```cs
[Browsable(true), Category("Category"),Description("just a test")]
public int Prop
{
    get { return _prop; }
    set { this._prop = value; }
}
```
but how to do it in Ironpython？  Looking forward to your reply

**Comment by slozier** [comment]
I'm not sure about this one, you might be able to do it with [clrtype](https://github.com/IronLanguages/ironpython3/blob/master/Src/IronPython/Lib/clrtype.py) with something like:
```py
import clrtype

class MyClass(metaclass=clrtype.ClrClass):
    __metaclass__ = clrtype.ClrClass # workaround for bug https://github.com/IronLanguages/ironpython3/issues/836
    
    from System.ComponentModel import CategoryAttribute
    Category = clrtype.attribute(CategoryAttribute)
    
    def __init__(self):
        self._prop = 0
    
    @property
    @Category("Category")
    def Prop(self):
        return self._prop
    
    @Prop.setter
    def Prop(self, value):
        self._prop = value
```

**Reply by hap1983** [reply]
thank for your apply， but  I encounter a new problems.  The error message is:  unbound method get_clr_type_name() must be called with MyClass instance as first argument (got nothing instead).   I am confused，could you  please help me to solve this problem?Thanks

**Reply by slozier** [reply]
I'm not sure, could you maybe post some sample code of what you're trying to run?

**Reply by hap1983** [reply]
OK， The following is my code:

```py
import System
import clr
import sys
import os

clr.AddReference('System.Windows.Forms')
clr.AddReference('System.Drawing')
clr.AddReference("System.Xml")
clr.AddReference("System.ComponentModel")

import System.Xml as xml
import System.Drawing
import System.Windows.Forms

import clrtype

import clrtype

class MyClass(clrtype.ClrClass):
    __metaclass__ = clrtype.ClrClass

    from System.ComponentModel import CategoryAttribute
    Category = clrtype.attribute(CategoryAttribute)

    def __init__(self):
        self._prop = 0

    @property
    @Category("Category")
    def Prop(self):
        return self._prop

    @Prop.setter
    def Prop(self, value):
        self._prop = value

class PropertyGridForm(Form):
    def __init__(self):
        self.InitializeComponent()
        self._obj = MyClass()
        self._propertyGrid1.SelectedObject = self._obj

    def InitializeComponent(self):
        self._propertyGrid1 = System.Windows.Forms.PropertyGrid()
        self.SuspendLayout()
        # 
        # propertyGrid1
        # 
        self._propertyGrid1.Dock = System.Windows.Forms.DockStyle.Fill
        self._propertyGrid1.Location = System.Drawing.Point(0, 0)
        self._propertyGrid1.Name = "propertyGrid1"
        self._propertyGrid1.Size = System.Drawing.Size(377, 409)
        self._propertyGrid1.TabIndex = 0
        # 
        # MainForm
        # 
        self.ClientSize = System.Drawing.Size(377, 409)
        self.Controls.Add(self._propertyGrid1)
        self.Name = "MainForm"
        self.Text = "PropertyGrid"
        self.ResumeLayout(False)

form = PropertyGridForm()
form.ShowDialog()
```

**Reply by slozier** [reply]
I see where the failure is occurring, this is incorrect:
```py
class MyClass(clrtype.ClrClass):
    __metaclass__ = clrtype.ClrClass
```
in my example I had:
```py
class MyClass(metaclass=clrtype.ClrClass):
    __metaclass__ = clrtype.ClrClass
```

and if you're using IronPython 2:
```py
class MyClass(object):
    __metaclass__ = clrtype.ClrClass
```

**Reply by hap1983** [reply]
Thanks for reply.  I use Ironpython 2.7, if I use “metaclass=clrtype.ClrClass”，there will be a error:   "unexpected token '=' at line 1065 in file ".  So I changed the code.   
Now  I  use "object" instead,  the interpreter give me  another error： Duplicate type name within an assembly，ArgumentException.  When I delete this line of code：“__metaclass__ = clrtype.ClrClass”，the error disappear and the code runs successfully，but the properties are not categorized in propertygrid control.  Could you  please give me an example of a successful operation with propetygrid in ironpython？I really appreciate your help

**Reply by slozier** [reply]
Unfortunately no luck getting it to work. Doesn't seem to be picking up on the Category attributes. I assume clrtype isn't adding the attribute to the .NET class as I was hoping it would.
